### PR TITLE
add explanation for E0429 (`self` use declaration must use brace syntax)

### PR DESCRIPTION
**Pull request by zackmdavis**
This is an item under #32777.

r? @GuillaumeGomez

### CODE REVIEW

File: src/librustc_resolve/diagnostics.rs [hunk 1]
```diff
@@ -1029,6 +1029,32 @@ struct Bar2; // ok!
 ```
 "##,
 
+E0429: r##"
+To import a namespace itself in addition to some of its members, the `self`
```
Comment (GuillaumeGomez) [inline src/librustc_resolve/diagnostics.rs]: This paragraph should just be the description of the error, not how to fix it. Put this paragraph just after the erroneous code example and replace it here by the error description.

Comment (zackmdavis) [inline src/librustc_resolve/diagnostics.rs]: Right. [Better?](https://github.com/rust-lang/rust/pull/33955/commits/06ebda837e50634cd1a73d54ce36a6fdec9688c5)

**Comment by GuillaumeGomez** [comment]
I liked the original explanation, it was more complete than the actual one. Could you re-use it or expand the current one please?

**Comment by zackmdavis** [comment]
@GuillaumeGomez 

> I liked the original explanation, it was more complete than the actual one. Could you re-use it

Unfortunately, I'm not sure I understand what you mean; both my [initial submission](https://github.com/zackmdavis/rust/blob/a2b7dfa3493b146936547b0f4ca44cc9cd57dd95/src/librustc_resolve/diagnostics.rs#L1032-L1056) and [the revision](https://github.com/zackmdavis/rust/blob/06ebda837e50634cd1a73d54ce36a6fdec9688c5/src/librustc_resolve/diagnostics.rs#L1032-L1054) would seem to cover the same information:

* `self` can't be used alone (second sentence as of a2b7dfa3, first sentence as of 06ebda83)
* to import a namespace as well as some members, `self` can be used in braces (first and last sentences as of a2b7dfa3; sentence after the erroneous snippet as of 06ebda83)
* if you just want to use the namespace, you don't need to bother with `self` (first sentence after the erroneous snippet as of a2b7dfa3; last sentence as of 06ebda83)

What made the first one (a2b7dfa3) feel more complete to you? 

In that, initial, draft, my outline was "start with a context-setting 'topic sentence' explaining what this brace-syntax feature is about, then explain the error, then point out how to just use the namespace itself, then have an example of the typical `::{self, OtherImport}` usage". After you pointed out that we want to start with a description of the error itself (without excessive contextualization), it seemed more parsimonious to combine the explanation of the brace syntax with the sentence introducing the typical-usage-of-brace-syntax snippet.

**Comment by GuillaumeGomez** [comment]
Hum... I see your point. Then I have nothing to add. Thanks!

@bors: r+ rollup

**Pull request merged**